src/react/NewPipelinePage.js

```diff
@@ -16,145 +16,159 @@
    limitations under the License.
 */
 
-import React from "react";
+import React, { useCallback, useMemo, useState } from "react";
 import {
   Button,
   Col,
   Form,
-  Row
+  Row,
+  Spinner
 } from "react-bootstrap";
+import Ajv from 'ajv';
 
 import SingleCardPage from "./ui/SingleCardPage.js";
 
-export default class NewPipelinePage extends React.Component {
-  constructor(props) {
-    super(props);
-
-    this.astarte = this.props.astarte;
-
-    this.addAlert = this.addAlert.bind(this);
-    this.closeAlert = this.closeAlert.bind(this);
-    this.onNameChanged = this.onNameChanged.bind(this);
-    this.onSourceChanged = this.onSourceChanged.bind(this);
-    this.onDescriptionChanged = this.onDescriptionChanged.bind(this);
-    this.createPipeline = this.createPipeline.bind(this);
-
-    this.state = {
-      alerts: new Map(),
-      alertId: 0,
-      pipelineName: "",
-      pipelineSource: "",
-      pipelineDescription: ""
-    };
-  }
-
-  addAlert(message) {
-    this.setState((state) => {
-      const newAlertId = state.alertId + 1;
-      let newAlerts = state.alerts;
-      newAlerts.set(newAlertId, message);
-
-      return Object.assign(state, {
-        alertId: newAlertId,
-        alerts: newAlerts
+let alertId = 0;
+const ajv = new Ajv({schemaId: 'id'});
+ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-04.json'));
+
+export default ({ astarte, history, blockId }) => {
+  const [alerts, setAlerts] = useState(new Map());
+  const [pipelineName, setPipelineName] = useState("");
+  const [pipelineSource, setPipelineSource] = useState("");
+  const [pipelineSchema, setPipelineSchema] = useState("");
+  const [pipelineDescription, setPipelineDescription] = useState("");
```

Comment:
Maybe we could decide, for all form pages, how to keep the state. For instance, just defining a single variable could help clean up the code:

```js
const [pipeline, setPipeline] = useState({
    name: "",
    source: "",
    schema: "",
    description: ""
  });
```

```jsx
<Form.Control
    value={pipeline.name}
    onChange={(e) => setPipeline({ ...pipeline, name: e.target.value})}
>
```

Reply:
If you do that, you lose control on the single component changes. In this specific example, you will rebuild the schema object every time the `pipeline` changes, even if only the name/description changed.

Reply:
Disregard previous comment, on further inspection, the dependency array can contain any value. Even a subset of the used object.

Reply:
Yes, exactly. If we actually had to extract a Form component, it is common to see in libraries handling forms that a single object is passed to the form to inject/hydrate form's input values. So keeping an entity's state together is not a new pattern. I'll leave this as point open for future discussion